packages/next/src/client/components/react-dev-overlay/internal/helpers/get-source-map-from-file.ts

```diff
@@ -0,0 +1,57 @@
+import fs from 'fs/promises'
+import path from 'path'
+import type { RawSourceMap } from 'next/dist/compiled/source-map08'
+import dataUriToBuffer from 'next/dist/compiled/data-uri-to-buffer'
+import { getSourceMapUrl } from './get-source-map-url'
+
+export async function getSourceMapFromFile(
+  filename: string
+): Promise<RawSourceMap | undefined> {
+  const fileContents = await fs
+    .readFile(filename, 'utf-8')
+    .catch(() => undefined)
+
+  if (!fileContents) {
+    return undefined
+  }
+
+  const sourceUrl = getSourceMapUrl(fileContents)
+
+  if (!sourceUrl) {
+    return undefined
+  }
+
+  if (sourceUrl.startsWith('data:')) {
+    let buffer
+
+    try {
+      buffer = dataUriToBuffer(sourceUrl)
+    } catch (err) {
+      console.error('Failed to parse source map URL:', err)
+      return undefined
+    }
+
+    if (buffer.type !== 'application/json') {
+      console.error(`Unknown source map type: ${buffer.typeFull}.`)
+      return undefined
+    }
+
+    try {
+      return JSON.parse(buffer.toString())
+    } catch {
+      console.error('Failed to parse source map.')
+      return undefined
+    }
+  }
+
+  const sourceMapFilename = path.resolve(path.dirname(filename), sourceUrl)
+
+  try {
+    const sourceMapContents = await fs.readFile(sourceMapFilename, 'utf-8')
+
+    return JSON.parse(sourceMapContents.toString())
+  } catch {
+    console.error('Failed to parse source map.')
```

Comment:
Feels like we should throw from this function and let the caller handle it. Then you also don't need the `.catch(() => undefined)` at L12.

At least having the cause of this error available somewhere would be helpful. It could've errored on parse or reading but we never know with this pattern